### PR TITLE
Refactor SinkServiceV1, PartitionBuffer and RecordService

### PR DESCRIPTION
**Pull request by sfc-gh-japatel**
- Just a Refactoring PR benefitted for StreamingIngest Based implementations.
- Also added some javadocs for more readability.
- No functional change.

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/PartitionBuffer.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * <p>Long lived would mean the data in partition would stay across two put APIs since the buffer
+ * thresholds were not met.
+ *
+ * <p>Please check respective implementation class for more details.
+ *
+ * @param <T> Return type of {@link #getData()}
+ */
+public abstract class PartitionBuffer<T> {
+  private int numOfRecord;
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/PartitionBuffer.java:26]: is numOfRecord = lastOffset-firstOffset? If yes, could we use it instead?

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM!

File: src/main/java/com/snowflake/kafka/connector/internal/PartitionBuffer.java (hunk 1 above)
Comment (sfc-gh-mnaides) [reply src/main/java/com/snowflake/kafka/connector/internal/PartitionBuffer.java:26]: If I recall correctly there can be gaps in offsets so we probably want to stick with numOfRecords

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 2]
```diff
@@ -1147,64 +1093,49 @@ public MetricRegistry getMetricRegistry() {
...
+     *
+     * <p>Please note {@link #insert(SinkRecord)} API is called from {@link
+     * com.snowflake.kafka.connector.SnowflakeSinkTask#put(Collection)} API and it is possible the
+     * buffered data is present across multiple PUT apis.
+     *
+     * <p>Check the usage of {@link #getData()} to understand when we would empty this buffer and
+     * when we would generate files in internal stage for snowpipe to ingest later using Snowpipe's
+     * REST APIs
+     */
+    private class SnowpipeBuffer extends PartitionBuffer<String> {
```
Comment (sfc-gh-mnaides) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:1107]: We should come up with how we want to differentiate between file based snowpipe and snowpipe streaming

**Review by sfc-gh-mnaides — APPROVED** [review]
LGTM, we might want to take another pass at how we're naming things to differentiate between Snowpipe as it is today and Snowpipe Streaming

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java (hunk 2 above)
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:1107]: Yes. We will defined a new PartitionBuffer Type. (More on that in this PR: https://github.com/snowflakedb/snowflake-kafka-connector/pull/382/files)

**Pull request merged**